source/solvers/cahn_hilliard.cc

```diff
@@ -112,7 +274,154 @@ template <int dim>
+
+  std::vector<double> local_phase_order_values(n_q_points);
+  std::vector<double> local_chemical_potential_values(n_q_points);
+
+  auto &exact_solution =
+    simulation_parameters.analytical_solution->cahn_hilliard;
+  exact_solution.set_time(simulation_control->get_current_time());
+
+  double l2_error_phase_order        = 0.;
+  double l2_error_chemical_potential = 0.;
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+    {
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit(cell);
+          fe_values[phase_order].get_function_values(evaluation_point,
+                                                     local_phase_order_values);
+          fe_values[chemical_potential].get_function_values(
+            evaluation_point, local_chemical_potential_values);
+
+          // Get the exact solution at all gauss points
+          exact_solution.vector_value_list(fe_values.get_quadrature_points(),
+                                           q_exactSol);
+
+          for (unsigned int q = 0; q < n_q_points; q++)
+            {
+              // Find the values of x and u_h (the finite element solution) at
+              // the quadrature points
+              double phase_order_sim   = local_phase_order_values[q];
+              double phase_order_exact = q_exactSol[q][0];
+
+              l2_error_phase_order += (phase_order_sim - phase_order_exact) *
+                                      (phase_order_sim - phase_order_exact) *
+                                      fe_values.JxW(q);
+
+              double chemical_potential_sim =
+                local_chemical_potential_values[q];
+              double chemical_potential_exact = q_exactSol[q][1];
+
+              l2_error_chemical_potential +=
+                (chemical_potential_sim - chemical_potential_exact) *
+                (chemical_potential_sim - chemical_potential_exact) *
+                fe_values.JxW(q);
+            }
+        }
+    }
+  l2_error_phase_order =
+    Utilities::MPI::sum(l2_error_phase_order, mpi_communicator);
+  l2_error_chemical_potential =
+    Utilities::MPI::sum(l2_error_chemical_potential, mpi_communicator);
+
+  return std::make_pair(std::sqrt(l2_error_phase_order),
+                        std::sqrt(l2_error_chemical_potential));
+}
+
+
+template <int dim>
+void
+CahnHilliard<dim>::calculate_phase_statistics()
+{
+  auto mpi_communicator = triangulation->get_communicator();
+
+  FEValues<dim> fe_values(*this->mapping,
+                          *fe,
+                          *this->cell_quadrature,
+                          update_values | update_gradients |
+                            update_quadrature_points | update_JxW_values);
+
+  const FEValuesExtractors::Scalar phase_order(0);
+
+  const unsigned int  n_q_points = cell_quadrature->size();
+  std::vector<double> local_phase_order_values(n_q_points);
+
+  double integral        = 0;
+  double max_phase_value = DBL_MIN;
```

Comment:
DBL_MIN is good, but an alternative you can use is std::numeric_limits<double>::min() and ::max(). It's useful when you want to distinguish between min() and lowest(), for example.

Reply:
agreed